raft/src/main/java/org/apache/kafka/raft/internals/BatchMemoryPool.java

```diff
@@ -96,9 +96,9 @@ public void release(ByteBuffer previouslyAllocated) {
         try {
             previouslyAllocated.clear();
 
-            if (previouslyAllocated.limit() != batchSize) {
+            if (previouslyAllocated.capacity() != batchSize) {
                 throw new IllegalArgumentException("Released buffer with unexpected size "
-                    + previouslyAllocated.limit());
+                    + previouslyAllocated.capacity());
```

Comment:
Although we cleared the buffer previously, the correct way to get the buffer size should be `Buffer#capacity`.